src/dist/replication/meta_server/duplication/meta_duplication_service.cpp

```diff
@@ -231,5 +231,106 @@ meta_duplication_service::new_dup_from_init(const std::string &remote_cluster_na
     return dup;
 }
 
+// ThreadPool(WRITE): THREAD_POOL_META_STATE
+void meta_duplication_service::recover_from_meta_state()
+{
+    ddebug_f("recovering duplication states from meta storage");
+
+    // /<app>/duplication/<dupid>/<partition_idx>
+    //                       |         |-> confirmed_decree
+    //                       |
+    //                       |-> json of dup info
+
+    for (const auto &kv : _state->_exist_apps) {
+        std::shared_ptr<app_state> app = kv.second;
+        if (app->status != app_status::AS_AVAILABLE) {
+            return;
+        }
+
+        _meta_svc->get_meta_storage()->get_children(
+            get_duplication_path(*app),
+            [this, app](bool node_exists, const std::vector<std::string> &dup_id_list) {
+                if (!node_exists) {
+                    // if there's no duplication
+                    return;
+                }
+                for (const std::string &raw_dup_id : dup_id_list) {
+                    dupid_t dup_id;
+                    if (!buf2int32(raw_dup_id, dup_id)) {
+                        // unlikely
+                        derror_f("invalid duplication path: {}",
+                                 get_duplication_path(*app, raw_dup_id));
+                        return;
+                    }
+                    do_restore_duplication(dup_id, app);
+                }
+            });
+    }
+}
+
+// ThreadPool(WRITE): THREAD_POOL_META_STATE
+void meta_duplication_service::do_restore_duplication_progress(
+    const duplication_info_s_ptr &dup, const std::shared_ptr<app_state> &app)
+{
+    for (int partition_idx = 0; partition_idx < app->partition_count; partition_idx++) {
+        std::string str_pidx = std::to_string(partition_idx);
+
+        // <app_path>/duplication/<dup_id>/<partition_index>
+        std::string partition_path = get_partition_path(dup, str_pidx);
+
+        _meta_svc->get_meta_storage()->get_data(
+            std::move(partition_path), [dup, partition_idx](const blob &value) {
+                if (value.size() == 0) {
```

Comment:
I guess that 'value.size() == 0' means that you can't find partition's duplication progress on zk. The type of 'value' is 'int64_t' indicating duplication progress.  I don't know if I understand it correct.

Reply:
right. I've added some comments here.